x-pack/plugins/apm/server/lib/services/get_services/get_services_items.ts

```diff
@@ -19,72 +20,315 @@ import {
 } from '../../helpers/setup_request';
 import { getServicesProjection } from '../../../../common/projections/services';
 
+const MAX_NUMBER_OF_SERVICES = 500;
+
 export type ServiceListAPIResponse = PromiseReturnType<typeof getServicesItems>;
-export async function getServicesItems(
-  setup: Setup & SetupTimeRange & SetupUIFilters
-) {
-  const { start, end, client } = setup;
 
-  const projection = getServicesProjection({ setup });
+const arrayUnionToCallable = <T extends any[]>(
+  array: T
+): Array<ValuesType<T>> => {
+  return array;
+};
+
+type ServicesItemsSetup = Setup & SetupTimeRange & SetupUIFilters;
+
+type ServicesItemsProjection = ReturnType<typeof getServicesProjection>;
+
+interface AggregationParams {
+  setup: ServicesItemsSetup;
+  projection: ServicesItemsProjection;
+}
 
-  const params = mergeProjection(projection, {
-    body: {
+const getTransactionDurationAvg = async ({
+  setup,
+  projection,
+}: AggregationParams) => {
+  const { client } = setup;
+
+  const response = await client.search(
+    mergeProjection(projection, {
       size: 0,
-      aggs: {
-        services: {
-          terms: {
-            ...projection.body.aggs.services.terms,
-            size: 500,
+      body: {
+        query: {
+          bool: {
+            filter: projection.body.query.bool.filter.concat({
+              term: {
+                [PROCESSOR_EVENT]: 'transaction',
+              },
+            }),
           },
-          aggs: {
-            avg: {
-              avg: { field: TRANSACTION_DURATION },
+        },
+        aggs: {
+          services: {
+            terms: {
+              ...projection.body.aggs.services.terms,
+              size: MAX_NUMBER_OF_SERVICES,
             },
-            agents: {
-              terms: { field: AGENT_NAME, size: 1 },
+            aggs: {
+              average: {
+                avg: {
+                  field: TRANSACTION_DURATION,
+                },
+              },
+            },
+          },
+        },
+      },
+    })
+  );
+
+  const { aggregations } = response;
+
+  if (!aggregations) {
+    return [];
+  }
+
+  return aggregations.services.buckets.map((bucket) => ({
+    name: bucket.key as string,
+    value: bucket.average.value,
+  }));
+};
+
+const getAgentName = async ({ setup, projection }: AggregationParams) => {
+  const response = await setup.client.search(
+    mergeProjection(projection, {
```

Comment:
Are we able to get rid of the projections if we remove the counts? I think that would make all queries 10x more readable.

Reply:
It would make it a little easier, as the contextual filters don't have to care about the aggregation needed to calculate a count. But there's a bunch of other stuff that a projection takes care of:
- range filter on `@timestamp`
- type-checking projection-specific parameters like `service.name` or `transaction.type`
- querying the appropriate index

I think the lightweight APM client you are proposing is trying to solve at least some of these things. I think we should also take latency metrics into account which will complicate things a bit.

Reply:
Projections creates an inversion of control which makes it very hard to see what's actually going on.

> range filter on @timestamp

We already have a helper for this?

> querying the appropriate index

We can make a helper for this, or like you say, we can use a lightweight client.

> type-checking projection-specific parameters like service.name or transaction.type

Not sure about what this is but it would be great if we can avoid IOC.